File: backend/app/routers/reports.py
from datetime import date
from typing import Literal
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Path, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import Principal, require_management
from app.schemas.report import (
    FeeOperationListResponse,
    FeeOperationResponse,
    FeePaidReceiptDetailResponse,
    FeePaidReceiptListResponse,
    PaymentReconciliationListResponse,
    PaymentReconciliationResolveRequest,
    PaymentReconciliationItemResponse,
)
from app.services.paid_report_service import (
    get_paid_fee_receipt,
    get_paid_fee_receipts,
)
from app.services.report_service import get_fee_operation, get_fee_operations
from app.services.payment_reconciliation_service import (
    list_payment_reconciliation,
    resolve_payment_reconciliation,
)

router = APIRouter(tags=["reports"])


@router.get("/fees/paid", response_model=FeePaidReceiptListResponse)
async def list_paid_fee_receipts(
    period: str | None = Query(default=None, pattern=r"^\d{4}-\d{2}$"),
    q: str | None = Query(default=None, min_length=1, max_length=100),
    date_from: date | None = None,
    date_to: date | None = None,
    payment_method: Literal["bank_transfer", "cash"] | None = None,
    payment_origin: Literal["manual", "manual_early", "pay2s"] | None = None,
    refund_state: Literal["NONE", "PARTIAL", "FULL", "REVERSED"] | None = None,
    cursor: str | None = Query(default=None, max_length=500),
    limit: int = Query(default=30, ge=10, le=100),
    db: AsyncSession = Depends(get_db),
    principal: Principal = Depends(require_management),
) -> FeePaidReceiptListResponse:
    return await get_paid_fee_receipts(
        db,
        period=period,
        query_text=q,
        date_from=date_from,
        date_to=date_to,
        payment_method=payment_method,
        payment_origin=payment_origin,
        refund_state=refund_state,
        cursor=cursor,
        limit=limit,
    )


@router.get("/fees/paid/{receipt_id}", response_model=FeePaidReceiptDetailResponse)
async def read_paid_fee_receipt(
    receipt_id: str = Path(min_length=20, max_length=500),
    db: AsyncSession = Depends(get_db),
    principal: Principal = Depends(require_management),
) -> FeePaidReceiptDetailResponse:
    receipt = await get_paid_fee_receipt(db, receipt_id)
    if receipt is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Không tìm thấy phiếu thu",
        )
    return receipt


@router.get("/fees/operations", response_model=FeeOperationListResponse)
async def list_fee_operations(
    action: Literal[
        "notify",
        "unnotify",
        "payment",
        "payment_reversal",
        "refund",
        "refund_reversal",
        "sync",
        "template_update",
    ]
    | None = None,
    period: str | None = Query(default=None, pattern=r"^\d{4}-\d{2}$"),
    q: str | None = Query(default=None, min_length=1, max_length=100),
    date_from: date | None = None,
    date_to: date | None = None,
    cursor: str | None = Query(default=None, max_length=300),
    limit: int = Query(default=30, ge=10, le=100),
    db: AsyncSession = Depends(get_db),
    principal: Principal = Depends(require_management),
) -> FeeOperationListResponse:
    return await get_fee_operations(
        db,
        action=action,
        period=period,
        query_text=q,
        date_from=date_from,
        date_to=date_to,
        cursor=cursor,
        limit=limit,
    )


@router.get("/fees/operations/{operation_id}", response_model=FeeOperationResponse)
async def read_fee_operation(
    operation_id: UUID,
    db: AsyncSession = Depends(get_db),
    principal: Principal = Depends(require_management),
) -> FeeOperationResponse:
    operation = await get_fee_operation(db, operation_id)
    if operation is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Không tìm thấy hoạt động học phí",
        )
    return operation


@router.get(
    "/fees/reconciliation",
    response_model=PaymentReconciliationListResponse,
)
async def list_fee_reconciliation(
    queue_status: Literal["PENDING", "PROCESSING", "POSTED", "REVIEW", "DEAD"] = Query(
        default="REVIEW", alias="status"
    ),
    limit: int = Query(default=100, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    principal: Principal = Depends(require_management),
) -> PaymentReconciliationListResponse:
    return await list_payment_reconciliation(db, queue_status=queue_status, limit=limit)


@router.post(
    "/fees/reconciliation/{queue_id}/resolve",
    response_model=PaymentReconciliationItemResponse,
)
async def resolve_fee_reconciliation(
    queue_id: UUID,
    payload: PaymentReconciliationResolveRequest,
    db: AsyncSession = Depends(get_db),
    principal: Principal = Depends(require_management),
) -> PaymentReconciliationItemResponse:
    try:
        response = await resolve_payment_reconciliation(
            db,
            queue_id=queue_id,
            payload=payload,
            actor_id=principal.user_id,
        )
        await db.commit()
        return response
    except HTTPException:
        await db.rollback()
        raise

File: frontend/src/lib/api/reports.ts
import { apiClient } from "@/lib/api/client";
import {
  feeOperationListSchema,
  feeOperationSchema,
  feePaidReceiptDetailSchema,
  feePaidReceiptListSchema,
  paymentReconciliationItemSchema,
  paymentReconciliationListSchema,
} from "@/lib/schemas/reports";
import type {
  FeeOperation,
  FeeOperationAction,
  FeeOperationListResponse,
  FeePaidReceiptDetail,
  FeePaidReceiptListResponse,
  FeePaidReceiptRefundState,
  FeePaymentMethod,
  PaymentReconciliationItem,
  PaymentReconciliationListResponse,
  PaymentReconciliationResolveAction,
} from "@/lib/types";
import { normalizePaidReportSearch } from "@/lib/reports/paid-report-view-model";

export type FeePaidReceiptFilters = {
  period?: string;
  q?: string;
  date_from?: string;
  date_to?: string;
  payment_method?: FeePaymentMethod | "";
  payment_origin?: "manual" | "manual_early" | "pay2s" | "";
  refund_state?: FeePaidReceiptRefundState | "";
  cursor?: string;
  limit?: number;
};

export async function getFeePaidReceipts(
  filters: FeePaidReceiptFilters = {},
  signal?: AbortSignal,
): Promise<FeePaidReceiptListResponse> {
  const { data } = await apiClient.get<unknown>("/reports/fees/paid", {
    signal,
    params: {
      ...filters,
      period: filters.period || undefined,
      q: filters.q ? normalizePaidReportSearch(filters.q) || undefined : undefined,
      payment_method: filters.payment_method || undefined,
      payment_origin: filters.payment_origin || undefined,
      refund_state: filters.refund_state || undefined,
      cursor: filters.cursor || undefined,
    },
  });
  return feePaidReceiptListSchema.parse(data);
}

export async function getFeePaidReceipt(
  receiptId: string,
  signal?: AbortSignal,
): Promise<FeePaidReceiptDetail> {
  const { data } = await apiClient.get<unknown>(
    `/reports/fees/paid/${encodeURIComponent(receiptId)}`,
    { signal },
  );
  return feePaidReceiptDetailSchema.parse(data);
}

export type FeeOperationFilters = {
  action?: FeeOperationAction | "";
  period?: string;
  q?: string;
  date_from?: string;
  date_to?: string;
  cursor?: string;
  limit?: number;
};

export async function getFeeOperations(
  filters: FeeOperationFilters = {},
  signal?: AbortSignal,
): Promise<FeeOperationListResponse> {
  const { data } = await apiClient.get<unknown>("/reports/fees/operations", {
    signal,
    params: {
      ...filters,
      action: filters.action || undefined,
      period: filters.period || undefined,
      q: filters.q ? normalizePaidReportSearch(filters.q) || undefined : undefined,
      cursor: filters.cursor || undefined,
    },
  });
  return feeOperationListSchema.parse(data);
}

export async function getFeeOperation(
  operationId: string,
  signal?: AbortSignal,
): Promise<FeeOperation> {
  const { data } = await apiClient.get<unknown>(
    `/reports/fees/operations/${encodeURIComponent(operationId)}`,
    { signal },
  );
  return feeOperationSchema.parse(data);
}

export async function getPaymentReconciliation(
  status = "REVIEW",
  signal?: AbortSignal,
): Promise<PaymentReconciliationListResponse> {
  const { data } = await apiClient.get<unknown>("/reports/fees/reconciliation", {
    signal,
    params: { status, limit: 100 },
  });
  return paymentReconciliationListSchema.parse(data);
}

export async function resolvePaymentReconciliation(
  queueId: string,
  payload: {
    action: PaymentReconciliationResolveAction;
    payment_request_id?: string | null;
    reason: string;
  },
): Promise<PaymentReconciliationItem> {
  const { data } = await apiClient.post<unknown>(
    `/reports/fees/reconciliation/${encodeURIComponent(queueId)}/resolve`,
    payload,
  );
  return paymentReconciliationItemSchema.parse(data);
}
